File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        if not s[0] in ('(', '{', '[') or len(s) % 2 == 1:
            return False

        stack = []
        for char in s:
            if char in ('(', '{', '['):
                stack.append(char)
            else:
                try:
                    top = stack.pop()
                except IndexError:
                    return False
                if char == ')' and top == '(':
                    continue
                elif char == '}' and top == '{':
                    continue
                elif char == ']' and top == '[':
                    continue
                else:
                    return False
        return len(stack) == 0


# @lc code=end
s = Solution()
s.isValid('()[]{}')

File: 20.valid-parentheses.go
package leetcode

// isValid TODO
// 20. 有效的括号
// stack
func isValid(s string) bool {
	if len(s)%2 != 0 {
		return false
	}
	stack := make([]byte, 0, len(s)/2)
	for i := 0; i < len(s); i++ {
		b := s[i]
		switch b {
		case '[', '{', '(':
			stack = append(stack, b)
		case ']':
			if len(stack) == 0 || stack[len(stack)-1] != '[' {
				return false
			}
			stack = stack[:len(stack)-1]
		case '}':
			if len(stack) == 0 || stack[len(stack)-1] != '{' {
				return false
			}
			stack = stack[:len(stack)-1]
		case ')':
			if len(stack) == 0 || stack[len(stack)-1] != '(' {
				return false
			}
			stack = stack[:len(stack)-1]
		default:
		}
	}
	return len(stack) == 0
}
